### PR TITLE
Add G(n,m) random graph generator

### PR DESCRIPTION
**Pull request by MoAllabbad**
This adds two G(n,m) graph generators, one directed and one undirected.
The generated graph will be a random graph out of all the possible
graphs that are n nodes and m edges with max n*(n-1) edges for directed
graphs and n*(n-1)/2 for undirected graphs, avoiding self-loops and
multigraphs. The run time is O(n+m)
Fixes #174

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1607,6 +1607,157 @@ pub fn undirected_gnp_random_graph(
...
+                    let v_index = NodeIndex::new(v as usize);
+                    inner_graph.add_edge(u_index, v_index, py.None());
+                }
+            }
+        }
+    } else {
+        let mut created_edges: isize = 0;
+        while created_edges < num_edges {
+            let u = rng.gen_range(0, num_nodes);
+            let v = rng.gen_range(0, num_nodes);
```
Comment (mtreinish) [inline src/lib.rs:1669]: Based on reading the docs in rand we probably should change this to use a `Uniform` distribution. https://docs.rs/rand/0.7.3/rand/distributions/uniform/struct.Uniform.html (specifically https://docs.rs/rand/0.7.3/rand/trait.Rng.html#method.gen_range which recommends using uniform if sampling the same range repeatedly)

Something like:

```suggestion
        let mut created_edges: isize = 0;
        let between = Uniform::new(0, num_nodes);
        while created_edges < num_edges {
            let u = between.sample(&mut rng);
            let v = between.sample(&mut rng);
```

(although you'll also need to add:

```rust
use rand::distributions::{Distribution, Uniform};
```

I realize that you just copied this form what I did in Gnp, so this isn't a blocker. I'll push a PR to update Gnp soon, and I can include Gnm in that if this PR merges first.

**Review by mtreinish — APPROVED** [review]
This LGTM, thanks for taking on this enhancement.

Before we merge this I think we should do the `!.is_some()` -> `is_none()` change. The other comments inline are just a couple of idle thoughts I had, but not worth blocking over (or necessarily even doing in the case of usize vs isize for parameters)

File: src/lib.rs (hunk 1 above)
Comment (mtreinish) [reply src/lib.rs:1669]: See: https://github.com/Qiskit/retworkx/pull/178 I'll update that after this merges so don't worry about changing this here.

File: src/lib.rs [hunk 2]
```diff
@@ -1607,6 +1607,157 @@ pub fn undirected_gnp_random_graph(
...
+            let u_index = NodeIndex::new(u as usize);
+            let v_index = NodeIndex::new(v as usize);
+            // avoid self-loops and multi-graphs
+            if u != v && !inner_graph.find_edge(u_index, v_index).is_some() {
```
[the remark this answers is not in the document]
Comment (MoAllabbad) [reply src/lib.rs:1673]: I agree, this is clearer

File: src/lib.rs [hunk 3]
```diff
@@ -1607,6 +1607,157 @@ pub fn undirected_gnp_random_graph(
...
```
[the remark this answers is not in the document]
Comment (MoAllabbad) [reply src/lib.rs:1748]: Agree

File: src/lib.rs [hunk 4]
```diff
@@ -1607,6 +1607,157 @@ pub fn undirected_gnp_random_graph(
...
+    num_nodes: isize,
+    num_edges: isize,
+    seed: Option<u64>,
+) -> PyResult<digraph::PyDiGraph> {
+    if num_nodes <= 0 {
+        return Err(PyValueError::new_err("num_nodes must be > 0"));
+    }
+    if num_edges < 0 {
+        return Err(PyValueError::new_err("num_edges must be >= 0"));
+    }
```
[the remark this answers is not in the document]
Comment (MoAllabbad) [reply src/lib.rs:1643]: I think this is more descriptive, not just in rust but in terms of the error message in python which is more pythonic. So, I think we should keep it.

File: src/lib.rs (hunk 1 above)
Comment (MoAllabbad) [reply src/lib.rs:1669]: > See: #178 I'll update that after this merges so don't worry about changing this here.

Got it

**Pull request merged**